### ISSUE TITLE
WarmUpCosine Wrong Computation

### ISSUE THREAD

**Issue by yonigottesman** [tensorflow/similarity#299]
Hi, while implementing the dino paper in tf #108 I noticed the WarmUpCosine output looks different. I checked the original simCLR implementation and its consistent with the dino one. Check out the outputs:

## DINO
https://github.com/facebookresearch/dino/blob/main/utils.py#L187
~~~python
def cosine_scheduler(base_value, final_value, epochs, niter_per_ep, warmup_epochs=0, start_warmup_value=0):
    warmup_schedule = np.array([])
    warmup_iters = warmup_epochs * niter_per_ep
    if warmup_epochs > 0:
        warmup_schedule = np.linspace(start_warmup_value, base_value, warmup_iters)

    iters = np.arange(epochs * niter_per_ep - warmup_iters)
    schedule = final_value + 0.5 * (base_value - final_value) * (1 + np.cos(np.pi * iters / len(iters)))

    schedule = np.concatenate((warmup_schedule, schedule))
    assert len(schedule) == epochs * niter_per_ep
    return schedule

lr_schedule = cosine_scheduler(
    lr ,  
    min_lr,
    epochs, steps_per_epoch,
    warmup_epochs=10,
)
plt.plot(lr_schedule)
~~~
<img width="580" alt="image" src="https://user-images.githubusercontent.com/4004127/196050996-e900136b-d92d-4383-8707-0d738fb937c5.png">

## simclr
https://github.com/google-research/simclr/blob/master/tf2/model.py#L78
~~~python
class WarmUpAndCosineDecay(tf.keras.optimizers.schedules.LearningRateSchedule):
  """Applies a warmup schedule on a given learning rate decay schedule."""

  def __init__(self, base_learning_rate, name="None"):
    super(WarmUpAndCosineDecay, self).__init__()
    self.base_learning_rate = base_learning_rate

    self._name = name

  def __call__(self, step):
    with tf.name_scope(self._name or 'WarmUpAndCosineDecay'):
      scaled_lr=self.base_learning_rate
      
      learning_rate = (
          step / float(warmup_steps) * scaled_lr if warmup_steps else scaled_lr)

      # Cosine decay learning rate schedule

      # TODO(srbs): Cache this object.
      cosine_decay = tf.keras.experimental.CosineDecay(
          scaled_lr, total_steps - warmup_steps)
      learning_rate = tf.where(step < warmup_steps, learning_rate,
                               cosine_decay(step - warmup_steps))

      return learning_rate
  
c = WarmUpAndCosineDecay(lr)
plt.plot([c(i) for i in range(total_steps)])
~~~
<img width="573" alt="image" src="https://user-images.githubusercontent.com/4004127/196051355-c6ff528a-33c6-4c16-a3ff-66626b7e5a9b.png">

## tensorflow-similarity
~~~python
c = WarmUpCosine(lr,total_steps,warmup_steps,alpha=min_lr/lr)
plt.plot([c(i) for i in range(total_steps)])
~~~
<img width="577" alt="image" src="https://user-images.githubusercontent.com/4004127/196051401-bba63ecb-6a7e-4f7c-bd81-9f3cac08fc89.png">

## New Implementation
I suggest this implementation for this package:

~~~python
class WarmupCosineDecay(tf.keras.optimizers.schedules.LearningRateSchedule):
    def __init__(self, base_learning_rate, steps, warmup_steps, alpha=0, name=None):
        super().__init__()
        self.base_learning_rate = tf.convert_to_tensor(base_learning_rate, dtype=tf.float32)
        self.cosine_decay = tf.keras.experimental.CosineDecay(base_learning_rate, steps - warmup_steps, alpha)
        self.warmup_steps = tf.convert_to_tensor(warmup_steps, dtype=tf.float32)
        self.name = name

    def __call__(self, step):
        with tf.name_scope(self.name or "WarmupCosineDecay") as name:
            step = tf.cast(step, tf.float32)
            learning_rate = tf.cond(
                step < self.warmup_steps,
                lambda: tf.math.divide_no_nan(step, self.warmup_steps) * self.base_learning_rate,
                lambda: self.cosine_decay(step - self.warmup_steps),
                name=name,
            )
            return learning_rate
c = WarmupCosineDecay(lr,total_steps,warmup_steps,alpha=min_lr/lr)
plt.plot([c(i) for i in range(total_steps)])
~~~
<img width="580" alt="image" src="https://user-images.githubusercontent.com/4004127/196052433-58cfd154-71e3-45a0-828a-ea36e6efbe68.png">

@owenvallis what do you think?

**Comment by owenvallis** [comment]
Thanks for finding this @yonigottesman, and for the proposed solution. Can you add this as a PR to the developments branch? I'll cherry pick the changes into the master branch once we merge the PR.

**Comment by owenvallis** [comment]
@yonigottesman, I'm actually using this for some benchmark test at the moment. I added your changes to my local branch and I can merge the changes if that works for you. Let me know.

**Comment by yonigottesman** [comment]
yes thats cool. I wanted to double check but if its working for you then go ahead

**Comment by owenvallis** [comment]
Uploaded the new version to the development branch. @yonigottesman, let me know if this works for you and I'll cherry pick it into the master branch. Here are a few plots of the new output with various settings.

![first_warmup_cosine_decay](https://user-images.githubusercontent.com/916864/197426994-8cbd1b7f-abdd-4629-a4d7-91398485b1de.png)

![second_warmup_cosine_decay](https://user-images.githubusercontent.com/916864/197427005-e0a2fd22-71d7-4cd2-868b-bd855f24a284.png)

![third_warmup_cosine_decay](https://user-images.githubusercontent.com/916864/197427021-16d88a72-797d-41fb-896c-6b4acd696d9c.png)
